### PR TITLE
statistics: skip impossible values when estimating row count

### LINKED ISSUE #32834: avoid impossible values are added into estimated row count
## Enhancement
Currently, in our cardinality estimation logic, the impossible values like `MaxValue` and `MinNotNull` are possible to be added into row count. This will lead to high estimation errors sometimes and needs to be avoided.

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #32834   ref #32758

Problem Summary: Please see the issue and test cases for detail.

### What is changed and how it works?

Skip impossible values when estimating row count.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: statistics/histogram.go [hunk 1]
```diff
@@ -1211,7 +1211,7 @@ func (c *Column) GetColumnRowCount(sctx sessionctx.Context, ranges []*ranger.Ran
 		// `betweenRowCount` returns count for [l, h) range, we adjust cnt for boundaries here.
 		// Note that, `cnt` does not include null values, we need specially handle cases
 		// where null is the lower bound.
-		if rg.LowExclude && !lowVal.IsNull() {
+		if rg.LowExclude && !lowVal.IsNull() && lowVal.Kind() != types.KindMaxValue && lowVal.Kind() != types.KindMinNotNull {
```
Comment (qw4990) [inline statistics/histogram.go:1214]: Please add some comment to explain in what case the condition `lowVal.Kind() != types.KindMaxValue && lowVal.Kind() != types.KindMinNotNull` can be true.

File: statistics/histogram.go [hunk 2]
```diff
@@ -1221,7 +1221,7 @@ func (c *Column) GetColumnRowCount(sctx sessionctx.Context, ranges []*ranger.Ran
 		if !rg.LowExclude && lowVal.IsNull() {
 			cnt += float64(c.NullCount)
 		}
-		if !rg.HighExclude {
+		if !rg.HighExclude && highVal.Kind() != types.KindMaxValue && highVal.Kind() != types.KindMinNotNull {
```
Comment (qw4990) [inline statistics/histogram.go:1224]: ditto

File: statistics/histogram.go (hunk 1 above)
Comment (time-and-fate) [reply statistics/histogram.go:1244]: Updated.

File: statistics/histogram.go (hunk 2 above)
Comment (time-and-fate) [reply statistics/histogram.go:1255]: The comments are added above.

File: cmd/explaintest/t/imdbload.test [hunk 3]
```diff
@@ -310,3 +310,11 @@ explain select * from aka_title where kind_id = 5;
...
 explain select * from aka_title where kind_id > 7;
 trace plan target = 'estimation' select * from aka_title where kind_id > 7;
+
+-- Actual row count: 0
+explain select * from keyword where ((phonetic_code = 'R1652') and (keyword > 'ecg-monitor' and keyword < 'killers'));
+trace plan target = 'estimation' select * from keyword where ((phonetic_code = 'R1652') and (keyword > 'ecg-monitor' and keyword < 'killers'));
+
+-- Actual row count: 37928
+explain select * from cast_info where (nr_order is null) and (person_role_id = 2) and (note >= '(key set pa: Florida');
+trace plan target = 'estimation' select * from cast_info where (nr_order is null) and (person_role_id = 2) and (note >= '(key set pa: Florida');
```
Comment (time-and-fate) [inline cmd/explaintest/t/imdbload.test:320]: Changes here are irrelevant to this PR. My next PR will need these test results for comparison.

File: statistics/histogram.go [hunk 4]
```diff
@@ -1238,8 +1238,10 @@ func (c *Column) GetColumnRowCount(sctx sessionctx.Context, ranges []*ranger.Ran
 		cnt := c.BetweenRowCount(sctx, lowVal, highVal, lowEncoded, highEncoded)
 		// `betweenRowCount` returns count for [l, h) range, we adjust cnt for boundaries here.
 		// Note that, `cnt` does not include null values, we need specially handle cases
-		// where null is the lower bound.
-		if rg.LowExclude && !lowVal.IsNull() {
+		//   where null is the lower bound.
```
Comment (Reminiscent) [inline statistics/histogram.go:1241]: useless change

Comment (time-and-fate) [reply statistics/histogram.go:1241]: I added indentation to make it easier to read. 😂

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**